client/db/src/lib.rs

```diff
@@ -1118,6 +1141,84 @@ impl<Block: BlockT> Backend<Block> {
 		self.storage.clone()
 	}
 
+	/// Ensure that the gap between the canonicalized and
+	/// finalized block is filled by canonicalizing all the blocks
+	/// up to the finalized block.
+	///
+	/// The canonicalized block could be behind the finalized block
+	/// in the case of delayed pruning. This inconsistency could
+	/// cause the database to miss-behave if started without this
+	/// option.
+	fn startup_canonicalization_gap(
+		&self,
+		transaction: &mut Transaction<DbHash>,
+	) -> ClientResult<()> {
+		// Read the finalized block from the database.
+		let finalized_num =
+			match self.storage.db.get(columns::META, meta_keys::FINALIZED_BLOCK).and_then(
+				|block_key| {
+					self.storage
+						.db
+						.get(columns::HEADER, &block_key)
+						.map(|bytes| Block::Header::decode(&mut &bytes[..]).ok())
+				},
+			) {
+				Some(Some(header)) => (*header.number()).saturated_into::<u64>(),
+				_ => return Ok(()),
+			};
+
+		// The last canonicalized block is stored in the state-db meta section.
+		let canonicalized_num = match self
```

Comment:
`LAST_CANONICAL` is an implementation detail of statedb, that should not be exposed. It is better to use `state_db.best_canonical()` in `from_database` and pass the value here.